Source/DafnyCore/AST/Cloner.cs

```diff
@@ -112,18 +112,19 @@ public class Cloner {
       } else if (d is TraitDecl) {
         var dd = (TraitDecl)d;
         var tps = dd.TypeArgs.ConvertAll(CloneTypeParam);
-        var mm = dd.Members.ConvertAll(d => CloneMember(d, false));
+        var mm = dd.Members.ConvertAll(member => CloneMember(member, false));
         var cl = new TraitDecl(Range(dd.RangeToken), dd.NameNode.Clone(this), m, tps, mm, CloneAttributes(dd.Attributes), dd.IsRefining, dd.ParentTraits.ConvertAll(CloneType));
         return cl;
+      } else if (d is DefaultClassDecl) {
```

Comment:
I think it's better to have the clone methods in the classes themselves, by implementing the `ICloneable` interface and letting `CloneDeclaration` check for the presence of that, but this is fine too.

By moving the cloning into the classes, you can also use a cloning constructor that calls into the cloning constructor of the base class. For `DefaultClassDecl`, that is only copying members inherited from base classes, that would mean the clone constructor can be empty and only call the clone constructor from its base class.

Reply:
I've seen that you've done that with `Expression` and that looks nice. If a `Cloner` subclass needs to do something special with the cloning of a declaration (or expression, for that matter), how does it do that? Would it then have the `if (d is ...)` structure that `CloneDeclaration` currently has? In other words, is it just the standard cloning that would be done via the `IClonable` interface?

Reply:
Yes and yes. Usually such subclasses of `Cloner` only override the cloning behavior of particular AST nodes, so this doesn't bring us back to a giant `if ... else if.. etc`

Reply:
Okay, then I agree that would be a good change. But since you say you're fine with the current version, too, I would like to request that we do it in a future PR instead.

Reply:
OK !